contract_reader.py

```diff
@@ -115,11 +104,62 @@ def get_orderbook():
             "Read order: sellAmount %40d %7s -- buyAmount %40d %7s -- accountID %s"
             % (sell_amount, sell_token, buy_amount, buy_token, o['accountID']))
 
-        orders.append({'accountID': o['accountID'],
-                       'sellToken': sell_token,
-                       'buyToken': buy_token,
-                       'sellAmount': str(int(sell_amount)),
-                       'buyAmount': str(int(buy_amount))})
+        orders_decoded.append({'accountID': o['accountID'],
+                               'sellToken': sell_token,
+                               'buyToken': buy_token,
+                               'sellAmount': str(int(sell_amount)),
+                               'buyAmount': str(int(buy_amount)),
+                               'validFrom': o['validFrom'],
+                               'validUntil': o['validUntil']})
+
+    return orders_decoded
+
+
+def get_current_orderbook():
+    """Get all order data from smart contract.
+
+    Returns:
+        A list of all orders.
+
+    """
+    orders = get_orderbook()
+
+    # Skip orders that are no longer or not yet valid.
+    batch_id = get_current_batch_id()
+    orders = list(filter(lambda order: order['validUntil'] <
+                         batch_id or order['validFrom'] > batch_id, orders))
+    return orders
+
+
+def get_orderbook():
+    """Get all order data from smart contract.
+
+    Returns:
+        A list of all orders.
+
+    """
+    orders = []
+
+    # Get orders via paginated approach
+    page_size = 100
+    current_user = '0x0000000000000000000000000000000000000000'
+    current_offset = 0
+    last_page_size = page_size
+
+    while last_page_size == page_size:
+
+        orders_decoded_page = decode_orders(
+            contract.functions.getEncodedUsersPaginated(
+                Web3.toChecksumAddress(current_user),
+                current_offset, page_size).call())
+
+        orders += orders_decoded_page
+        if len(orders_decoded_page) > 0:
+            current_user = orders_decoded_page[-1]['accountID']
+            current_offset = len(
+                [o for o in orders if o['accountID'] == current_user])
+
```

Comment:
Isn't `current_offset = len(orders)` ?

Reply:
No, it is always an offset per user, due to the way, we are storing the orders per user in the smart contract